IdentityCore/src/cache/MSIDTokenCache.m

```diff
@@ -24,50 +24,386 @@
+        MSID_LOG_INFO_PII(context, @"No matching access token found.");
+        return YES;
+    }
+    
+    if ([matchedTokens[0] isExpired])
+    {
+        MSID_LOG_INFO(context, @"Access token found in cache is already expired.");
+        MSID_LOG_INFO_PII(context, @"Access token found in cache is already expired.");
+        return YES;
+    }
+    
+    *outAccessToken = matchedTokens[0];
+    return YES;
+}
+
+
+- (MSIDToken *)getRTforUser:(MSIDUser *)user
+                             authority:(NSURL *)authority
+                              clientId:(NSString *)clientId
+                               context:(id<MSIDRequestContext>)context
+                                 error:(NSError **)error
 {
+    // Look for new cache, with utid and uid
+    if (user.userIdentifier)
+    {
+        MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyForRefreshTokenWithUserId:user.userIdentifier
+                                                                        clientId:clientId
+                                                                     environment:authority.msidHostWithPortIfNecessary];
+        
+        MSIDToken *token = [_dataSource itemWithKey:key serializer:_jsonSerializer context:context error:error];
+        if (token)
+        {
+            return token;
+        }
+    }
+    
+    // Look for old cache
+    // if there is upn
+    if (user.upn)
+    {
+        MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyWithAuthority:authority clientId:clientId resource:nil upn:user.upn];
+        
+        MSIDToken *token = [_dataSource itemWithKey:key serializer:_keyedArchiverSerialize context:context error:error];
+        
+        if (token)
+        {
+            return token;
+        }
+    }
+    
     return nil;
 }
 
+- (NSArray<MSIDToken *> *)getAllRTsForClientId:(NSString *)clientId
+                                       context:(id<MSIDRequestContext>)context
+                                         error:(NSError **)error
+{
+    if (!clientId)
+    {
+        return nil;
+    }
+    
+    NSMutableArray<MSIDToken *> *allRTs = [NSMutableArray<MSIDToken *> new];
+    
+    // Look at new cache
+    MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyForRefreshTokenWithClientId:clientId];
+    
+    NSArray *newTokens = [_dataSource itemsWithKey:key serializer:_jsonSerializer context:context error:error];
+    if (!newTokens)
+    {
+        return nil;
+    }
+    
+    for (MSIDToken *token in newTokens)
+    {
+        if (token.tokenType == MSIDTokenTypeRefreshToken)
+        {
+            [allRTs addObject:token];
+        }
+    }
+    
+    // Look at old cache
+    NSArray *legacyTokens = [_dataSource itemsWithKey:[MSIDTokenCacheKey keyForAllItems]
+                                           serializer:_keyedArchiverSerialize
+                                              context:context
+                                                error:error];
+    
+    if (!legacyTokens)
+    {
+        return nil;
```

Comment:
We shouldn't return nil here, what if we found only new RT but nothing from old cache. We should return new tokens?

Reply:
I beg to differ here,
!legacyTokens here indicate some kind of caching error, not a successful-but-zero-items returned result.
In that regard, I still think we should return nil here even though there might be some token cache from new RT.

Reply:
Hmmm, what would happen if an app using MSAL freshly has no access to shared access group? Would that still return an empty array? I just want to differentiate between cases where it failed for an unexpected reason and not fail the whole call for something expected.

Reply:
I agree that we need to differentiate such cases, but datasource should determine error cases, and thus if it returns nil, I would think we should treat it as an error.

if an app using MSAL freshly has no access to shared access group>> I'd think that is an error, if I am understanding correctly :p